lib/universe.ts

```diff
@@ -0,0 +1,137 @@
+ * @param callbackFn - a function to execute for each element in the iterator
+ * @param concurrency - number of elements to apply the function at the same time. Default to 1
+ */
+export async function concurrentForEach<T>(
+	it: IterableIterator<T>,
+	callbackFn: (t: T) => PromiseLike<void>,
+	concurrency = 1,
+): Promise<void> {
+	const run = async () => {
+		const next = it.next();
+		if (next.value && !next.done) {
+			await callbackFn(next.value);
+			await run();
+		}
+	};
+	const runs = [];
+	for (let i = 0; i < concurrency; i++) {
+		runs.push(run());
+	}
+	await Promise.all(runs);
+}
+
+/**
+ * @summary recursively find all files under the directory that match the given filter
+ * @function
+ * @memberof module:universe
+ *
+ * @param dir - base directory to start the search
+ * @param filter - filtering function to indicate that a file should be selected
+ */
+async function findFiles(
+	dir: string,
+	filter: (filePath: string, stat: Stats) => boolean = () => true,
+): Promise<string[]> {
+	const dirFiles = await fs.readdir(dir);
+
+	const filePaths: string[] = [];
+	const dirPaths: string[] = [];
+	for (const fileName of dirFiles) {
+		const filePath = path.join(dir, fileName);
+
+		const stat = await fs.stat(filePath);
+		if (stat.isDirectory()) {
+			dirPaths.push(filePath);
+		} else if (filter(filePath, stat)) {
+			filePaths.push(filePath);
+		}
+	}
+
+	const allFiles = await Promise.all(dirPaths.map((d) => findFiles(d, filter)));
+	return filePaths.concat(...allFiles);
+}
+
+interface FromFsOptions {
+	/**
+	 * Additional filters to apply to json files when loading a universe from FS
+	 */
+	filter: (filePath: string, stat: Stats) => boolean;
+
+	/**
+	 * Only load the canonical version of the contract and ignore
+	 * aliases
+	 */
+	canonicalOnly: boolean;
+}
+
+export class Universe extends Contract {
+	constructor() {
+		super({ type: UNIVERSE });
+	}
+
+	/**
+	 * @summary recursively looks up all json files under a directory and adds them
+	 * to the universe
+	 * @function
+	 * @static
+	 * @name module:contrato.Universe.fromFs
+	 * @public
+	 *
+	 * @param dir full path of the directory to load
+	 * @param options additional configuration for the search and build process
+	 *
+	 */
+	static async fromFs(
+		dir: string,
+		{ filter = () => true, canonicalOnly = false }: Partial<FromFsOptions> = {},
+	): Promise<Universe> {
+		const allFiles = await findFiles(
+			dir,
+			(filePath, stat) =>
+				path.extname(filePath) === '.json' && filter(filePath, stat),
+		);
+
+		const universe = new Universe();
+		const children: Contract[] = [];
+		await concurrentForEach(
+			allFiles.values(),
+			async (file) => {
+				const contents = await fs.readFile(file, { encoding: 'utf8' });
+				let source = JSON.parse(contents);
+
+				if (canonicalOnly) {
+					// Ignore aliases
+					const { aliases, ...obj } = source;
+					source = obj;
+				}
+
+				children.push(...Contract.build(source));
+			},
+			10,
+		);
```

Comment:
You can use https://www.npmjs.com/package/p-map to handle the concurrency with an off-the-shelf package
```suggestion
		const children = (await pMap(
			allFiles,
			async (file) => {
				const contents = await fs.readFile(file, { encoding: 'utf8' });
				let source = JSON.parse(contents);

				if (canonicalOnly) {
					// Ignore aliases
					const { aliases, ...obj } = source;
					source = obj;
				}

				return Contract.build(source);
			},
			{ concurrency: 10 },
		).flat();
```

It also supports returning an iterable and can do backpressure on that iterable if necessary (which it might be in the context of contracts stuff?)

Reply:
I managed to import p-map on a new commit. We do use concurrency when generating base images, p-map will probably be useful for that too. I'll think about how p-map can make sense for the blueprint reproduce functionality